### DISCUSSION TITLE
Privacy

### DISCUSSION THREAD

**Discussion by DutchPete** [brianpetro/obsidian-smart-connections#78] [Q&A] [answered]
How's the privacy aspect? OpenAI gives you an API key that belongs to OpenAI, and Smart Connections processes your notes. Can OpenAI read your notes, do they get stored?

What does processing by Smart Connections involve?

**Comment by brianpetro** [comment, marked as answer]
@DutchPete, thanks for reaching out about this. Let's see if I can help clarify this relatively common question for you and others that may see this.

To begin, there is only one external API provider that Smart Connections uses: OpenAI's API. This contrasts with similar plugins that rely on other third-party servers for processing.

While I wish that we could all run AI locally on our computers, which would be ideal from a privacy perspective, instead, relying on only one external API greatly reduces the potential for possible privacy issues. 

Additionally, your embeddings/vectors are stored locally in a file on your computer. Once again, this contrasts with some other plugins that use third-party services for embedding/vector storage.

But, since the plugin does use OpenAI for the necessary AI components, which at this time there are no comparable alternatives, you will have to review their Terms of Service and other documentation to understand better what to expect when using their services. 

Lastly, to help with occasions where there are only some sensitive files/folders, Smart Connections provides exclusion settings that you can use to prevent certain files/folders from ever being processed by OpenAI.

I hope that clarifies how Smart Connections works and its scope.

Thanks for reaching out about this, 
Brian🌴

**Reply by pinuke** [reply]
Well, this is something I've also wanted to ask for a while. Do you think we could eventually support Hugging Face's pretrained AI's as Hugging Face is locally hosted? or even just support other Pretrained AI APIs like Facebook's OPT-175B?

**Reply by brianpetro** [reply]
Supporting additional models shouldn't be an issue once they're available.

Hugging Face provides a similar service to OpenAI, where you could send your data to Huggingface spaces instead of the OpenAI API. This implementation is already possible but not a priority since it doesn't solve the privacy issue.

Hugging Face does provide access to models for download locally. However, I haven't seen any evidence that this is worthwhile. Any attempts I've seen at running such models locally have been woefully underwhelming in their abilities and speed.

**Reply by pinuke** [reply]
> Hugging Face does provide access to models for download locally. However, I haven't seen any evidence that this is worthwhile. Any attempts I've seen at running such models locally have been woefully underwhelming in their abilities and speed.

Oh for sure. Locally hosted infrastructure will never compare to the behemoths of Microsoft and Facebook, but it would be the only way to achieve 100% privacy. I've considered eventually hosting some of Hugging Face's bigger models whenever I can afford the hardware lol (may be several years before I do that).

**Reply by brianpetro** [reply]
Yeah, unfortunately, consumer hardware is far from being capable at the moment. The people trying are using >$5K mac books, and the results are still unimpressive.

**Reply by pinuke** [reply]
Well, to be fair, mac books don't really have an impressive amount of VRAM, and most of the time the limiting factor on AI performance is going to be VRAM.

If this was going to be implemented, it would definitely need to rely on consumer-end mining rigs and definitely not traditional consumer-end PCs lol.

**Reply by spolisar** [reply]
> Hugging Face does provide access to models for download locally. However, I haven't seen any evidence that this is worthwhile. Any attempts I've seen at running such models locally have been woefully underwhelming in their abilities and speed.

I think you're a bit off in dismissing locally hosted options. Sure, nothing else is on the same level as ChatGPT for cases like conversation, but there is genuine competition in embedding performance. If you want to see how different models stack up against each other in embedding, take a look at [MTEB](https://huggingface.co/spaces/mteb/leaderboard).  There are models that outperform the GPT embedding model (text-embedding-ada-002) that can be run on consumer-end PCs. For some of them, a dedicated GPU might be necessary, but there are others that can come close on average and even beat GPT embedding on some tasks.

**Reply by brianpetro** [reply]
Hey @spolisar , welcome to the discussion!

And thank you for the helpful resource with the latest benchmarks.

I'm in no way trying to dismiss the importance of locally hosted options. Undoubtedly, they are the future.

I'd be happy to integrate as soon as:
1) the benchmarks check out on reasonable consumer hardware, and
2) the model can be installed relatively simply

It would be helpful if you could keep us updated on the potential local models as they begin to meet the above criteria.

Thanks for the insights!
Brian🌴

**Reply by spolisar** [reply]
Thanks for the welcome.

What do you think is reasonable consumer hardware? Are dedicated GPUs on the table, or just CPUs? It's obvious that an A100 isn't reasonable for a consumer, but where do you draw the line?

Would you count something like a docker image as a relatively simple installation?

**Reply by pinuke** [reply]
@spolisar reasonable would be anything the average Obsidian user has. You have to remember, Obsidian is equal parts developer community to writing and academics. Unless there is an interest in gaming/dev, the average college student is probably not going to have more than a consumer-grade laptop (maybe not even more than a chromebook)

**Comment by DutchPete** [comment]
Thanks for this detailed answer, Brian. I'll check out OpenAI's ToS.

**Reply by pinuke** [reply]
Thanks for the contribution and its good that you show that you want to learn more about the plugin before using it!

Just make sure to put replies to an answer as a reply and not answer. Github discussions work slightly differently than issues!
